### PR TITLE
Create Blog “why-is-redfish®-different-from-other-rest-apis-part-1”

### PR DESCRIPTION
**Pull request by donzef**
Automatically generated by Netlify CMS

### CODE REVIEW

File: content/blog/why-is-redfish®-different-from-other-rest-apis-part-1.md [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+authorimage: /img/fdz-photoprofile.png
+thumbnailimage: ""
+disable: false
+tags:
+  - Redfish
+  - ilo-restful-api
+---
+## Redfish® brief recap
+
+Redfish® is a low level management RESTful API standardized by the Distributed Management Task Force (<a href="https://dmtf.org" target="_blank">DMTF</a>) consortium. DMTF publishes several standards related to server management, like the Common Information Model (CIM) implemented as OpenPegasus on Linux and Windows Management Instrumentation (WMI) on Microsoft Windows.
```
Comment (drensing) [inline content/blog/why-is-redfish®-different-from-other-rest-apis-part-1.md:14]: place a hyphen between "low" and "level" so it's "low-level"

File: content/blog/why-is-redfish®-different-from-other-rest-apis-part-1.md [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+
+Redfish® is born in August 2015. As of today, more than <a href="https://dmtf.org/about/list" target="_blank">sixty members</a> contribute to its development: computer makers, storage and network device manufacturers and software developers. A funny story says that "Redfish®" comes from the name of a restaurant near the Compaq drive in Houston, Texas, where initiators of the project were meeting before DMTF took over the entire project.
```
Comment (drensing) [inline content/blog/why-is-redfish®-different-from-other-rest-apis-part-1.md:16]: If you keep this statement as is, you still need to change "is" to "was" (past tense). However, I think you might actually want to reword it to something like:
"Redfish made its first appearance in August 2015." or "Redfish was first introduced in August 2015."

Change "A funny story says that" to "A funny story relates how "Redfish" comes from..."

File: content/blog/why-is-redfish®-different-from-other-rest-apis-part-1.md [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+
+The main goal of Redfish® is to replace the Intelligent Platform Management Interface (<a href="https://www.intel.com/content/www/us/en/products/docs/servers/ipmi/ipmi-home.html" target="_blank">IMPI</a>) originally created by Intel®. In a nutshell, Redfish® is able to monitor, configure and perform actions (i.e. power-on/off) on remote (out-of-band) or local (in-band) servers. Also, it provides an event subscription mechanism that can replace the Simple Network Management Protocol (SNMP).
```
Comment (drensing) [inline content/blog/why-is-redfish®-different-from-other-rest-apis-part-1.md:18]: Modify the first sentence to read:
"The main **reason Redfish was introduced was** to replace the..."

File: content/blog/why-is-redfish®-different-from-other-rest-apis-part-1.md [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+
+The adoption of this standard across the industry has been very quick and this article is an attempt to explain some of the specificities that have contributed to this success.
```
Comment (drensing) [inline content/blog/why-is-redfish®-different-from-other-rest-apis-part-1.md:20]: Change "this success" to "its success"

File: content/blog/why-is-redfish®-different-from-other-rest-apis-part-1.md [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+
+## Redfish® specificities
+
+Redfish® is classified as a RESTful API as all the fundamentals are present, like a service/client model, HTTPS JSON formatted requests transferring representational states from and to identified end points. However, several specificities are unique, like the separation of the protocol and the data model and the possibility to enhance the data model with Original Equipment Manufacturer (OEM) extensions.
```
Comment (drensing) [inline content/blog/why-is-redfish®-different-from-other-rest-apis-part-1.md:24]: Shouldn't it be a client/server model?

Where you have the second comma in the first sentence, replace it with the word "and" so it reads:
"...all the fundamentals are present, like a client/server model and HTTPS JSON Formatted requests that transfer representational states from and to identified end points."

File: content/blog/why-is-redfish®-different-from-other-rest-apis-part-1.md [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+
+### Service-client architecture
+
+The Redfish® service is included in the server management controller firmware and cannot be separated from it. In this article, I will name this controller “BMC”, for “Baseboard Management Controller”. Every modern server worthy of the name has a BMC which starts as soon as an electrical cable is plugged into a power supply. Internally, the BMC communicates with most server subsystems (BIOS, network controllers, storage, power supplies, fans, etc.). It can communicate with the operating system as well, either through a proprietary protocol or a <a href="https://www.dmtf.org/sites/default/files/standards/documents/DSP0270_1.0.0.pdf" target="_blank">host interface</a>. Externally, the BMC includes a network connector which allows it to communicate with other network services (DHCP, SMTP, SNMP, etc.). At HPE, this BMC is called “HPE iLO”, at AMI it is OpenBMC, at Dell iDRAC…
```
Comment (drensing) [inline content/blog/why-is-redfish®-different-from-other-rest-apis-part-1.md:28]: In the 3rd sentence, change the word "which" to "that" so it reads "...has a BMC that starts as soon..."

Also change the "which" in the second to last sentence to "that"

https://www.grammarly.com/blog/which-vs-that/

File: content/blog/why-is-redfish®-different-from-other-rest-apis-part-1.md [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+
+Proprietary monitoring applications (i.e., <a href="https://www.hpe.com/us/en/hpe-greenlake-compute-ops-management.html" target="_blank">HPE Compute Ops Management</a>) or open source applications (<a href="https://github.com/ODIM-Project/ODIM" target="_blank">ODIM</a>, <a href="https://github.com/nsfcac/Nagios-Redfish-API-Integration" target="_blank">Nagios</a>, etc.) constitute native Redfish® clients or can become so using specific plugins.
+
+### Remote or local management
+
+In most cases, Redfish® clients performs <a href="https://en.wikipedia.org/wiki/Out-of-band_management" target="_blank">out-of-band management</a>. However, it is entirely possible to install the client in the OS of the server to be managed and to access the BMC via an internal path to the server (in-band-management). This path can be proprietary like HPE's CHIF (<a href="https://developer.hpe.com/blog/chif-driver-not-found/" target="_blank">Channel Interface</a>) or the <a href="https://en.wikipedia.org/wiki/Ethernet_over_USB" target="_blank">Ethernet over USB</a> protocol which allows an IP connection between the OS and the BMC.
```
Comment (drensing) [inline content/blog/why-is-redfish®-different-from-other-rest-apis-part-1.md:38]: Change "performs" to "perform" (subject/verb agreement).

Place a comma before "which" in the last sentence.

File: content/blog/why-is-redfish®-different-from-other-rest-apis-part-1.md [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+
+### Protocol and data model separation
+
+The Redfish® protocol specification is published in DMTF document <a href="https://www.dmtf.org/dsp/DSP0266" target="_blank">DSP0266</a> while data modeling in <a href="https://www.dmtf.org/dsp/DSP0268" target="_blank">DSP0268</a>. This separation allows flexibility with regard to the management of version numbers. Without this separation, a minor change to the protocol specifications (e.g. adding an acronym to the glossary) would involve an increment of the overall specification version number and impose a set of unnecessary tests to requalify clients.
```
Comment (drensing) [inline content/blog/why-is-redfish®-different-from-other-rest-apis-part-1.md:42]: I think you're missing a word in the first sentence. It also needs a comma. I think it should read "...document DSP0266, while data modeling **is in** DSP0268."

File: content/blog/why-is-redfish®-different-from-other-rest-apis-part-1.md [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+
+#### What’s inside the protocol specification?
+
+* Boring but necessary topics like the normative framework (definitions, symbols, glossary, typographical conventions, etc.).
+* Things a little more interesting like compliance with other standards (<a href="https://www.odata.org/" target="_blank">Open Data</a>, <a href="https://swagger.io/specification/" target="_blank">Swagger/OpenAPI</a>, etc.).
+* We also find very practical things like the ways to authenticate (basic authentication, session token, etc.) and the endpoints where authentication is necessary (everywhere except at the root `/redfish/v1`).
```
Comment (drensing) [inline content/blog/why-is-redfish®-different-from-other-rest-apis-part-1.md:48]: Modify this to "Some very practical things like the ways...

File: content/blog/why-is-redfish®-different-from-other-rest-apis-part-1.md [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+* Surprising things like receiving an "error" messages for successful requests. Figure 2 shows the successful erasure of a storage logical volume (LUN) whose response contains the `error{}` object with a message explaining the success of the operation! This type of incongruous response is a consequence of compliance with the <a href="https://www.odata.org/" target="_blank">Open Data</a> standard.
```
Comment (drensing) [inline content/blog/why-is-redfish®-different-from-other-rest-apis-part-1.md:49]: Change "messages" to "message". (since you placed "an" in front of "error", you make it singular)

File: content/blog/why-is-redfish®-different-from-other-rest-apis-part-1.md [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+
+![Figure 2: The error contains a success message !](/img/fig2-successerror.png "Figure 2: The error contains a success message !")
+
+#### Data model
+
+Responses to Redfish® requests consist of JSON packets containing key/value properties defined by the DMTF in a schema file. This name of schema files is contained in the `@odata.type` property that must be present in each response.
```
Comment (drensing) [inline content/blog/why-is-redfish®-different-from-other-rest-apis-part-1.md:55]: I don't quite understand what you mean by "This name of scheme files". Maybe it should be "file"?

File: content/blog/why-is-redfish®-different-from-other-rest-apis-part-1.md [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+
+For example, the schema defining the root of the Redfish® tree is `#ServiceRoot`. Its full name returned by `curl -sk https://bmc-ip/redfish/v1 | jq '."@odata.type"'` is: `#ServiceRoot.v1_13_0.ServiceRoot`. Appended to the `#ServiceRoot` fragment, a version number (`1_13_0`) and then a subtype which, in this specific case, is identical to the main schema. All schemas are publicly available on the <a href="at https://redfish.dmtf.org/schemas" target="_blank">DMTF website</a> and, are sometimes included in the service itself (see the [Self Sufficient Model](#self-sufficient-model) paragraph below). Note that schema versions can evolve independently of each other.
```
Comment (drensing) [inline content/blog/why-is-redfish®-different-from-other-rest-apis-part-1.md:57]: In the third sentence, change "which, in this specific case..." to "that, in this specific case..."

File: content/blog/why-is-redfish®-different-from-other-rest-apis-part-1.md [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+
+With a close look to the Redfish® root diagram in (Figure 3), you will notice the presence of endpoints allowing access to the modeling of the server subsystems. For example, the `Chassis{}` object points to `/redfish/v1/Chassis` which contains the exhaustive collection of the URIs (`./{ChassisId}`) modeling the different chassis constituting the server (racks, blades, enclosures, storage enclosures, etc.).
```
Comment (drensing) [inline content/blog/why-is-redfish®-different-from-other-rest-apis-part-1.md:59]: Place a comma before "which contains the exhaustive..."
Remove "the" prior to "URIs"

File: content/blog/why-is-redfish®-different-from-other-rest-apis-part-1.md [hunk 14]
```diff
@@ -0,0 +1,134 @@
...
+
+![Figure 3: source dmtf.org/education](/img/fig3-resourcemap.png "Figure 3: source dmtf.org/education")
+
+### OEM extensions
+
+Computer makers have the possibility of modeling proprietary properties (not standardized) and constituting added value compared to the competition. For example, HPE iLO can store firmware updates before deploying them to their respective components. This specificity is described in the proprietary schema `#HpeiLOUpdateServiceExt` present in the `Oem.Hpe{}` object under the URI `/redfish/v1/UpdateService` (Figure 4).
```
Comment (drensing) [inline content/blog/why-is-redfish®-different-from-other-rest-apis-part-1.md:65]: I'd reword this first sentence to:
"Computer makers are able to model proprietary properties (not standardized), which constitutes added value compared to the competition."

File: content/blog/why-is-redfish®-different-from-other-rest-apis-part-1.md [hunk 15]
```diff
@@ -0,0 +1,134 @@
...
+
+## Resource collections
+
+As shown in figure 3 above, links contained in the Redfish root tree, do not provide direct access to the properties of the server subsystems. Instead, they point to “resource collections”. The collection members are described by a unique schema. For example, the response to the GET `/redfish/v1/Chassis` request made against the BMC of an HPE Superdome consisting of 2 chassis in a rack, is a collection containing four members: RackGroup, Rack1, r001u03b and r001u08b (Figure 5). These four members have the `#Chassis` diagram in common.
+
+![Figure 5: HPE Superdome chassis collection](/img/fig5-sdfchassis.png "Figure 5: HPE Superdome chassis collection")
+
+Another example: An HPE ProLiant server containing a storage enclosure connected to a (“modern”) storage controller is represented by two members in the chassis collection. A member pointing to the chassis containing the motherboard of the server (`/redfish/v1/Chassis/1`) and a member pointing to the backplane of the storage enclosure (`/redfish/v1/Chassis/DE040000`).
```
Comment (drensing) [inline content/blog/why-is-redfish®-different-from-other-rest-apis-part-1.md:77]: The last sentence in this paragraph is not a complete sentence. Consider rephrasing it or appending it to the previous sentence by changing the period after "the chassis collection" to a semi-colon ";" and changing "A member" to "a member"

File: content/blog/why-is-redfish®-different-from-other-rest-apis-part-1.md [hunk 16]
```diff
@@ -0,0 +1,134 @@
...
+
+Each member of the collection is an endpoint for accessing the member properties. Links to related resources constitute transversal access to other subsystems. For example, the storage chassis in Figure 6 contains a cross-link to the storage device under the "Systems" tree (`/redfish/v1/Systems/1/Storage/DE040000`).
+
+![Figure 6: Universal backplane model](/img/fig6-storagechassis.png "Figure 6: Universal backplane model")
+
+### Naming convention of collection members
+
+The naming of collection members is specific to each implementation of the Redfish® service. The BMC endpoint for an HPE server of type ProLiant is: `/redfish/v1/Managers/1`. That of an OpenBMC is: `/redfish/v1/bmc` and that of a Superdome is `/redfish/v1/Managers/RMC`. Dell, SuperMicro, Lenovo probably have their own way of naming their BMCs, while remaining compliant with the specifications.
```
Comment (drensing) [inline content/blog/why-is-redfish®-different-from-other-rest-apis-part-1.md:85]: Modify the second sentence to read:
"The BMC endpoint for an HPE ProLiant server is:..."

File: content/blog/why-is-redfish®-different-from-other-rest-apis-part-1.md [hunk 17]
```diff
@@ -0,0 +1,134 @@
...
+
+The naming flexibility given to Redfish® services generates extra work for client developers: their code has to discover the names of collection members before being able to access their properties. If they assume a specific URI, their code will fail when used against another computer maker or after a firmware update.
```
Comment (drensing) [inline content/blog/why-is-redfish®-different-from-other-rest-apis-part-1.md:87]: oh! well, that's a problem, isn't it... when it fails after a firmware update!

File: content/blog/why-is-redfish®-different-from-other-rest-apis-part-1.md [hunk 18]
```diff
@@ -0,0 +1,134 @@
...
+
+## Registries
+
+The data modeling described above don't represent resources that have interdependencies (i.e., BIOS attribute %1 depends of attribute %2) as well as information or error messages and their arguments (e.g.: "error on disk %s in location %s"). To address this problem, Redfish® uses "registries".
+
+Figure 7 is extracted from a BIOS registry which has the effect of prohibiting the activation of Windows secure mode support if the server does not contain a TPM (Trusted Platform Module). It is interesting to note that the BIOS registry also describes the GUI menus (`MapToProperty=GrayOut`).
```
Comment (drensing) [inline content/blog/why-is-redfish®-different-from-other-rest-apis-part-1.md:93]: Change "which" to "that"

File: content/blog/why-is-redfish®-different-from-other-rest-apis-part-1.md [hunk 19]
```diff
@@ -0,0 +1,134 @@
...
+Redfish® is considered "self-sufficient" because all (or almost) of the information regarding data modeling, operations and possible actions is documented and programmatically accessible. Client programmers no longer have to consult paper specifications before implementing them. They can transfer this task to their code in order to ensure a certain portability in time and space.
+
+### Accessing Schemas and registries
+
+Redfish® provides access to schemas and registries through the following endpoints:
+
+* `/redfish/v1/JsonSchemas`
+* `/redfish/v1/Registries`
+
+These URIs contain links to all schemas and registries used by the service. They point to documents stored in the BMC, if it has the storage capacity, or to the official DMTF website. In the later case, Redfish® clients must have a mean to access the Internet to download those documents. They are be helpful to identify property interdependencies, supported values or read-write capability.
```
Comment (drensing) [inline content/blog/why-is-redfish®-different-from-other-rest-apis-part-1.md:112]: In the third sentence, I think you mean "In the latter case,..." And you might mean "must have a means to access the Internet"

In the final sentence, it should probably read:
"They are helpful to identify..." (don't need the word "be")

File: content/blog/why-is-redfish®-different-from-other-rest-apis-part-1.md [hunk 20]
```diff
@@ -0,0 +1,134 @@
...
+
+### Allowed requests
+
+The Redfish® protocol requires GET request responses to contain the `Allow` header indicating the possible operations on the URI. Figure 9 shows that the URI `/redfish/v1/Chassis/{ChassisId}` allows GET, HEAD, and PATCH operations against this URI.
+
+![Figure 9: Possible operations on the URI](/img/fig9-allowheader.png "Figure 9: Possible operations on the URI")
+
+After verification, the client code can change (PATCH) the properties marked as `ReadOnly=False` in the schema. The main server chassis in Figure 9 shows the PATCH method and the IndicatorLED property can be modified (Figure 10), so we should be able to turn on the small blue LED on this chassis to better identify it in the data center. A good programmer will perform all of these checks before sending their changes. The bad programmer will not perform any checks and will leave the end user to deal with the error(s) returned by the service!
```
Comment (drensing) [inline content/blog/why-is-redfish®-different-from-other-rest-apis-part-1.md:120]: Change "we" to "one" where it reads "(Figure 10), so one should be able to perform..."

File: content/blog/why-is-redfish®-different-from-other-rest-apis-part-1.md [hunk 21]
```diff
@@ -0,0 +1,134 @@
...
+
+### Deprecated resources
+
+Over time, certain properties are deprecated. This information is mentioned in the schemas. Thus, clients receiving a “property not found” response will be able to consult the schema and check if by chance this property has not been deprecated. This type of verification will avoid comments from the programmer like "I don’t understand, it worked before!". But before what? Figure 10 shows an excerpt from the `#Chassis` schema, version 1.22.0, stating that the `IndicatorLED` property has been deprecated in version 1.14.0.
```
Comment (drensing) [inline content/blog/why-is-redfish®-different-from-other-rest-apis-part-1.md:124]: lol... "I don't understand, it worked before!" (love it!)

File: content/blog/why-is-redfish®-different-from-other-rest-apis-part-1.md [hunk 22]
```diff
@@ -0,0 +1,134 @@
...
+
+![Figure 10: Deprecated property](/img/fig10-propertydeprecated.png "Figure 10: Deprecated property")
+
+## What else?
+
+The separation of protocol from data modeling, collections, OEM extensions and registries form the basis of the Redfish® standard. Access to schemas and registries as well as the list of possible operations for each URI are important features allowing the writing of portable client code in time and space.
+
+However, Redfish® is full of other particularities, conceptual and practical, which are very interesting to study. For example the mode of communication with additional storage or network cards, updating firmware, or even the use of Swordfish® schemas developed by the Storage and Network Industry Association (<a href="https://www.snia.org/" target="_blank">SNIA</a>). Many security aspects are also addressed by Redfish® such as the integrity of the various server components. Read the second part of this Redfish® introduction for a complete view of this standard API.
```
Comment (drensing) [inline content/blog/why-is-redfish®-different-from-other-rest-apis-part-1.md:132]: Hmmm... that second sentence isn't a complete sentence... in spite of how long it is.  You need a subject and a verb...

**Review by drensing — COMMENTED** [review]
Lovely blog post, Francois!

**Review by choukden — CHANGES_REQUESTED** [review]
Hi Francois,
Sorry for the delay for review of your part 1 of the series.
This looks great to me. I learned a lot about RedFish!
Just one minor comment about the blog post you are referring in the post abut Hypermedia API concept.

Let us know once you have took into account all the suggestions for edits.

We will wait for the second part review and readiness to publish the series.

Thanks
Denis
